bothan-htx/src/api/websocket.rs

```diff
@@ -0,0 +1,301 @@
+use std::io::Read;
+
+use flate2::read::GzDecoder;
+use futures_util::stream::{SplitSink, SplitStream};
+use futures_util::{SinkExt, StreamExt};
+use serde_json::json;
+use tokio::net::TcpStream;
+use tokio_tungstenite::tungstenite::error::Error as TungsteniteError;
+use tokio_tungstenite::tungstenite::http::StatusCode;
+use tokio_tungstenite::tungstenite::Message;
+use tokio_tungstenite::{connect_async, MaybeTlsStream, WebSocketStream};
+use tracing::warn;
+
+use crate::api::error::{ConnectionError, MessageError, SendError};
+use crate::api::types::{HtxResponse, Pong};
+
+/// A connector for establishing a WebSocket connection to the Htx API.
+pub struct HtxWebSocketConnector {
+    url: String,
+}
+
+impl HtxWebSocketConnector {
+    /// Creates a new instance of `HtxWebSocketConnector`.
+    pub fn new(url: impl Into<String>) -> Self {
+        Self { url: url.into() }
+    }
+
+    /// Connects to the Htx WebSocket API.
+    pub async fn connect(&self) -> Result<HtxWebSocketConnection, ConnectionError> {
+        let (wss, resp) = connect_async(self.url.clone()).await?;
+
+        let status = resp.status();
+        if StatusCode::is_server_error(&status) || StatusCode::is_client_error(&status) {
+            warn!("failed to connect with response code {}", resp.status());
+            return Err(ConnectionError::UnsuccessfulHttpResponse(resp.status()));
+        }
```

Comment:
why http on websocket?

Reply:
its weird but tungstenite connect_async function really return resp with http response ?????

Reply:
may we change error message and name but inner type remain the same?

Reply:
done